### PR TITLE
docs: add Kafka ADRs

### PR DESCRIPTION
**Pull request by robrap**
**Notes:**

* This is in preparation of an update to OEP-52, which will point to a variety of other ADRs as appropriate.
* These ADRs are all marked "Provisional".
* Also added commit to fix `make docs`.

**Changes:**

* add ADR for Kafka-based event (moved from OEP-52)
* add ADR for Confluent hosting (moved from OEP-52/branch)
* added new README.rst to discuss temporary location
  of these ADRs.
* Renamed top-level README (added .rst)
* Added `/decisions` to ADR path.
* Re-numbered/small edits to consumer ADR.

### CODE REVIEW

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+------
+
+Provisional
+
+Context
+-------
+
+The draft `OEP-52: Event Bus Architecture`_ explains how the Open edX platform would benefit from an event bus, as well as providing some additional decisions around the event bus. One decision is to enable the event bus technology to be pluggable through some abstraction layer.
+
+However, an initial technology, even if multiple could eventually be selected, is still required.
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:15]: The construction of this sentence is a little hard to follow

```suggestion
This still requires selecting specific technology for the first implementation.
```

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+.. _`OEP-52: Event Bus Architecture`: https://github.com/openedx/open-edx-proposals/pull/233
+
+Decision
+--------
+
+An initial implementation of an event bus for the Open edX platform will be implemented using `Kafka`_. This implementation will be used by edx.org, and available to the Open edX community.
+
+However, given the intention of creating an abstraction layer around the event bus technology, this decision does not preclude the introduction of alternative event bus implementations based on other technologies in the future.
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:24]: I think the first clause unnecessarily complicates the sentence since it's not really a contradiction.

```suggestion
This decision does not preclude the introduction of alternative event bus implementations based on other technologies in the future.
```

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+^^^^
+
+* Battle-tested, widely adopted, big community, lots of documentation and answers.
+* Enables event replay-ability.
+
+Cons
+^^^^
+
+* Complex to manage, including likely manual scaling.
+* Simple consumers requires additional code for some messaging features.
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:58]: s/requires/require/

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+Consequences
+------------
+
+* Operators will need to deploy and manage the selected infrastructure, which is likely to be complex. If Apache Kafka is selected, there are likely to be a set of auxiliary parts to provide all required functionality for our message bus. However, third-party hosting is also available (see separate decision).
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:63]: Nit: 'which will likely be' is a little more compact a construction than 'which is likely to be'

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+* Most of the consequences of an event bus should related to `OEP-52: Event Bus Architecture`_ more generally, and hopefully will not be Kafka specific.
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:64]: s/related/relate/

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+Rejected Alternatives
+---------------------
+
+Apache Pulsar
+~~~~~~~~~~~~~
+
+Although rejected to start, `Apache Pulsar`_ remains an option for those looking for an alternative to Kafka.
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:72]:
```suggestion
Although rejected for initial edx.org implementation, `Apache Pulsar`_ remains an option for those looking for an alternative to Kafka.
```

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+Pros
+^^^^
+
+* Ease of scalability (built-in, according to docs).
+* Ease of data retention capabilities.
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:78]: 'ease of capabilities' is a strange construction. Maybe  `good data retention capabilities` or `easy data retention`

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+.. _Apache Pulsar: https://pulsar.apache.org/
+.. _comparisons and myths of Kafka vs Pulsar: https://dzone.com/articles/pulsar-vs-kafka-comparison-and-myths-explored
+
+Redis
+~~~~~
+
+Pros
+^^^^
+
+* Already part of Open edX platform
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:99]: Nit: I don't like using periods for sentence fragments but if we use them everywhere else we might as well be consistent : P
```suggestion
* Already part of Open edX platform.
```

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+Cons
+^^^^
+
+* Cannot be shared as an open source solution.
+* Events are not replayable.
+
+Additional References
+---------------------
+
+* Technology comparisons performed by edX.org
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:137]: Nit
```suggestion
* Technology comparisons performed by edX.org:
```

File: edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+* The 2U/edX team responsible for this new infrastructure does not have the capacity to manage this in house.
+
+.. _Kafka: https://kafka.apache.org/
+
+Decision
+--------
+
+.. note::
+
+    This decision is specific to edx.org and is not a requirement or recommendation of how other deployments in the Open edX community manage their brokers. However, the background may still be useful to other commity members.
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst:24]:
```suggestion
    This decision is specific to edx.org and is not a requirement or recommendation for how other deployments in the Open edX community manage their brokers. However, the background may still be useful to other commity members.
```

File: edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+`Amazon MSK`_ is an AWS managed service that supplies the Apache Kafka core platform only.
+
+The `Confluent Platform`_ adds additional capabilities, some of which are only commercially available.
+
+  * `Schema Registry <https://www.confluent.io/product/confluent-platform/data-compatibility/>`__
+  * Monitoring and alerting capabilities (Commercial)
+  * Self-balancing clusters (Commercial)
+  * Tiered storage (Commercial) (future feature of Apache Kafka)
+  * Infinite retention (Cloud only?)
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst:41]: Can you clarify why this is a question?

**Review by rgraber — COMMENTED** [review]
Pretty much all comments are just syntax and sentence construction nits. No objection to the content except one request for clarification

File: edx_arch_experiments/kafka_consumer/docs/decisions/0002_kafka_consumer_command.rst [hunk 12]
```diff
@@ -1,15 +1,14 @@
...
-======
-
-In Progress
+------
+Provisional
 
 Context
-=======
+-------
 As outlined in the upcoming OEP-52, edX.org has elected to go with Apache Kafka as our event bus implementation. Though the decision presented here is predicated on this particular edX.org decision, it is included to help other Open edX users evaluate Kafka for their own purposes. The standard pattern for consuming events with Kafka is to poll in a loop and process messages as they come in. According to the Confluent team it is a best practice to limit each consumer to a single topic (Confluent is a platform for industry-scale Kafka management)::
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0002_kafka_consumer_command.rst:10]: Are we going to keep the Kafka decision in OEP-52 or will that just talk about the general architecture after your revisions?

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst (hunk 1 above)
Comment (robrap) [reply edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:15]: Thanks. I'll use this, but add `a` before `specific technology`.

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst (hunk 2 above)
Comment (robrap) [reply edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:24]: 👍

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst (hunk 3 above)
Comment (robrap) [reply edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:58]: 👍

File: edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst (hunk 11 above)
Comment (robrap) [reply edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst:41]: Because I'm not sure if it is part of the Confluent Platform, or just part of Confluent Cloud, both of which are Commercial. I am asking Confluent and maybe I can firm up an answer.

File: edx_arch_experiments/kafka_consumer/docs/decisions/0002_kafka_consumer_command.rst (hunk 12 above)
Comment (robrap) [reply edx_arch_experiments/kafka_consumer/docs/decisions/0002_kafka_consumer_command.rst:10]: Good catch. Although the OEP will reference the ADR, I updated this text to reference the ADR explicitly instead.

**Comment by robrap** [comment]
@rgraber: Your feedback has been addressed. Thanks!

File: edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst (hunk 11 above)
Comment (robrap) [reply edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst]: I have updated.

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+.. _Kafka: https://kafka.apache.org/
+
+Why Kafka?
+~~~~~~~~~~
+
+Kafka is a distributed streaming platform. Kafka's implementation maps nicely to the pub/sub pattern. However, some native features of a message broker are not built-in.
+
+Kafka has been around for a long time. See `Thoughtworks's technology radar introduced Kafka`_ as "Assess" in 2015, and "Trial" in 2016. It never moved up to "Adopt", and also never moved down to "Hold". Read `Thoughtwork's Kafka decoder page`_ to learn more about its benefits and trade-offs, and how it is used.
+
+More recently, the `Thoughtworks's technology radar introduced Apache Pulsar`_ as "assess" in 2020, and the `technology radar introduced Kafka API without Kafka`_ in 2021. This both demonstrates the de facto standard of the Kafka API, but also Thoughtwork's hope to find a less complex alternative.
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:35]: First sentence is missing a verb.

File: edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+* The 2U/edX team responsible for this new infrastructure does not have the capacity to manage this in house.
+
+.. _Kafka: https://kafka.apache.org/
+
+Decision
+--------
+
+.. note::
+
+    This decision is specific to edx.org and is not a requirement or recommendation for how other deployments in the Open edX community manage their brokers. However, the background may still be useful to other commity members.
```
Comment (rgraber) [inline edx_arch_experiments/kafka_consumer/docs/decisions/0003_kafka_managed_hosting.rst:24]:
```suggestion
    This decision is specific to edx.org and is not a requirement or recommendation for how other deployments in the Open edX community manage their brokers. However, the background may still be useful to other community members.
```

**Review by rgraber — APPROVED** [review]
2 typos/nits, otherwise LGTM!

File: edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst (hunk 13 above)
Comment (robrap) [reply edx_arch_experiments/kafka_consumer/docs/decisions/0001_kafka_based_event_bus.rst:35]: Do you mean the sentence beginning with "More recently"? The verb is "introduced", and appears twice, although both times in the midst of a link, which may be why it was missed?
